---
title: Pipeline Report
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
params:
  report: "report"
  conf: "conf"
  dat: "dat"
  name: "filename"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

<style type="text/css">
#TOC {
  left: 40px;
  margin: 20px 0px 25px 0px;
}

.main-container {
  max-width: 2400px;
  margin-left: 100px;
  margin-right: auto;
}

.imgTooltip {
      display: none;
}

.ItemsTooltip:hover .imgTooltip {
      display: block;
      position: absolute;
      z-index: 1;
}
</style>

# Report Overview

This page contains an overview for the analyses that have been carried out. Links are provided to each per-trait specific analysis in the dataset tables. All tables and figures are downloadable and all separate report pages are self-contained html files, meaning these can be sent to other people without extra files.

Details and explanations of analyses and results are given throughout. Certain terms related to MR are hyperlinked to the MR Dictionary, which offers an explanation for that term.

# Overview of Results

This shows a basic overview of all of the results, which are given in more detail below and in individual reports. This overview shows a basic "Pass" or "Fail" mark if that exposure-outcome pair passes or fails that analysis. Passes are defined for each of the analyses:
<ul>
  <li> MR: P value < `r conf$mr_p_thres`.
  <li> Steiger filtering: P < 0.05 and correctly estimated direction of effect.
  <li> Colocalisation: H4 > `r conf$coloc_strong_thres` is strong evidence, H4 > `r conf$coloc_moderate_thres` is moderate evidence, and weak evidence otherwise.
</ul>

```{r overview_res}
mrres <- params$report$results %>%
  dplyr::select(id.exposure, id.outcome, exposure, outcome, pval) %>%
  dplyr::mutate(mrflag = ifelse(pval < conf$mr_p_thres, "Pass", "Fail"))

steigerres <- params$report$steigerresults %>%
  dplyr::select(id.exposure, id.outcome, flag) %>%
  dplyr::mutate(flag = ifelse(flag == "True", "Pass",
                              ifelse(flag == "Unknown", "Uncertain", "Fail")))

colocres <- params$report$cresults %>%
  dplyr::select(name1, name2, nsnps, H4) %>%
  dplyr::rename(exposure = name1, outcome = name2) %>%
  dplyr::mutate(colocflag = ifelse(H4 > conf$coloc_strong_thres & nsnps > 50, "Strong",
                                   ifelse(H4 > conf$coloc_moderate_thres & nsnps > 50, "Moderate", "Weak"))) %>%
  dplyr::mutate(colocflag = ifelse(colocflag == "", "NA", colocflag))

#dgidbres <- params$reports$dgidbresults %>%
#  dplyr::select(id, Druggable.Genome, Clinically.Actionable, Drug.Resistant) %>%
#  dplyr::rename(Drug.Genome = Druggable.Genome,
#                Clin.Act = Clinically.Actionable) %>%
#  dplyr::mutate(Drug.Genome = ifelse(Drug.Genome == T, "Y", "N"),
#                Clin.Act = ifelse(Clin.Act == T, "Y", "N"),
#                Drug.Resistant = ifelse(Drug.Resistant == T, "Y", "N"))

dplyr::left_join(mrres, steigerres, by = c("id.exposure", "id.outcome")) %>%
  dplyr::left_join(colocres, by = c("exposure", "outcome")) %>%
#  dplyr::left_join(dgidbres, by = c("id.exposure" = "id")) %>%
  dplyr::select(-pval, -nsnps, -H4) %>%
  dplyr::rename(Exposure.ID = id.exposure,
                Outcome.ID = id.outcome,
                Exposure = exposure,
                Outcome = outcome,
                MR.Pass = mrflag,
                Steiger.Pass = flag,
                Coloc.Pass = colocflag) %>%
  dplyr::mutate(Exposure = paste0("<a href='./traits/", Exposure, ".html'>", Exposure, "</a>")) %>%
  dplyr::mutate(Outcome = paste0("<a href='./traits/", Outcome, ".html'>", Outcome, "</a>")) %>%
  dplyr::relocate(Exposure, Outcome,
                  Exposure.ID, Outcome.ID,
                  MR.Pass, Steiger.Pass, Coloc.Pass) %>%
  DT::datatable(rownames = F,
                escape = F,
                caption = "Overview of all results",
                filter = "top",
                extensions = c("Buttons"),
                options = list(dom = "BDfrtip", 
                               buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download')))) %>%
  DT::formatStyle(
    "MR.Pass",
    color = DT::styleEqual(c("Pass", "Fail"),
                           c("darkgreen", "red")),
    fontWeight = DT::styleEqual("Pass", "bold")
  ) %>%
  DT::formatStyle(
    "Steiger.Pass",
    color = DT::styleEqual(c("Pass", "Uncertain", "Fail"),
                           c("darkgreen", "orange", "red")),
    fontWeight = DT::styleEqual("Pass", "bold")
  ) %>%
  DT::formatStyle(
    "Coloc.Pass",
    color = DT::styleEqual(c("Strong", "Moderate", "Weak", "NA"),
                           c("darkgreen", "orange", "red", "red")),
    fontWeight = DT::styleEqual("Strong", "bold")
  )# %>% 
#  DT::formatStyle(
#    c("Drug.Genome", "Clin.Act", "Drug.Resistant"),
#    color = DT::styleEqual(c("Y", "N"),
#                           c("darkgreen", "red")),
#    fontWeight = DT::styleEqual("Y", "bold")
#  )
```

# Overview of Drug Target Evidence

## Open Targets Platform

```{r ot-table}
brks <- seq(from = 0, to = 1, by = 0.1)
clrs <- round(seq(180, 40, length.out = length(brks) + 1), 0) %>%
  {paste0("rgb(", ., ",", ., ",180)")}

to_table <- params$report$otresults %>%
  dplyr::select(-id1, -id2) %>%
  dplyr::relocate(name2, .after = name1) %>%
  dplyr::rename(Exposure = name1,
                Outcome = name2,
                Overall.Score = overall.ot,
                Literature = literature.ot,
                RNA.Expression = rna_expression.ot,
                Genetic.Assoc = genetic_assoc.ot,
                Somatic.Mutations = somatic_mute.ot,
                Known.Drugs = known_drug.ot,
                Animal.Model = animal_model.ot,
                Affected.Pathways = affected_pathway.ot)

to_table %>%
  DT::datatable(rownames = F,
                escape = F,
                filter = 'top',
                extensions = c("Buttons"),
                options = list(order = list(2, "desc"), 
                               dom = "Bfrtip", 
                               buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download')))) %>%
    DT::formatSignif(columns = c("Overall.Score", 
                               "Literature", 
                               "RNA.Expression", 
                               "Genetic.Assoc",
                               "Somatic.Mutations",
                               "Known.Drugs",
                               "Animal.Model",
                               "Affected.Pathways"), 
                   digits = 3) %>%
  DT::formatStyle(names(to_table),
                  background = DT::styleColorBar(range(0, 1), 'lightblue'),
                  backgroundSize = '98% 88%',
                  backgroundRepeat = 'no-repeat',
                  backgroundPosition = 'center')
```

## Drug-Genome Interaction DB

```{r dgidb-table}
params$report$dgidbresults %>%
  dplyr::select(-id) %>%
  dplyr::rename(Trait = trait) %>%
  dplyr::mutate(Druggable.Genome = ifelse(Druggable.Genome == T, "Yes", "No"),
                Clinically.Actionable = ifelse(Clinically.Actionable == T, "Yes", "No"),
                Drug.Resistant = ifelse(Drug.Resistant == T, "Yes", "No")) %>%
  DT::datatable(rownames = F,
                escape = F,
                filter = 'top',
                extensions = c("Buttons"),
                options = list(order = list(0, "asc"),
                               dom = "Bfrtip",
                               buttons = list ('copy',
                                               list(extend = 'collection',
                                                    buttons = c('csv', 'excel'),
                                                    text = 'Download')))) %>%
  DT::formatStyle(
    c("Druggable.Genome", "Clinically.Actionable", "Drug.Resistant"),
    color = DT::styleEqual(c("Yes", "No"),
                           c("darkgreen", "red")),
    fontWeight = DT::styleEqual("Yes", "bold")
  )
```

## Possible Drugs

```{r dgidb-drugs}
params$report$dgidbdrugs %>%
  dplyr::select(-id) %>%
  dplyr::rename(Trait = trait) %>%
  DT::datatable(rownames = F,
                filter = 'top',
                extensions = c("Buttons"),
                options = list(dom = "Bfrtip",
                               buttons = list ('copy',
                                               list(extend = 'collection',
                                                    buttons = c('csv', 'excel'),
                                                    text = 'Download'))))
```

# Datasets

Tables presented here contain a breakdown and overview of the datasets. Trait.Name links take you to the specific trait results page and the Trait.ID links will take you to the OpenGWAS DB report, if applicable for that trait.

## Exposure datasets

There were `r nrow(unique(params$report$exposures))` exposure datasets provided. Details are provided in the following table for each dataset, including how many SNPs were removed at each stage of the QC pipeline.

```{r exposures}
params$report$exposures %>%
  dplyr::mutate(Trait.Annotated = paste0("<a href='./traits/", Trait.Annotated, ".html'>", Trait.Annotated, "</a>"),
                Trait.ID = ifelse(From.IEUGWASDB == 1, paste0("<a href='https://gwas.mrcieu.ac.uk/datasets/", Trait.ID, "/'>", Trait.ID, "</a>"), Trait.ID)) %>%
  dplyr::relocate(Trait.ID, .after = Trait.Annotated) %>%
  dplyr::select(-From.IEUGWASDB, -Trait.Name) %>%
  DT::datatable(rownames = F,
                escape = F,
                filter = 'top',
                extensions = 'Buttons',
                caption = "Breakdown for each exposure dataset.",
                options = list(dom = 'Bfrtip',
                               buttons = list('copy',
                                              list(extend = 'collection',
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download'))))

```
**Column headers and meanings**
<ul>
  <li><code>Trait.Annotated</code>: Name of the trait - Links to that trait's results report.</li>
  <li><code>Trait.ID</code>: Trait ID - Links to the OpenGWAS DB, if applicable.</li>
  <li><code>SNPs</code>: Number of SNPs initially loaded from the GWAS (either from the file or from the OpenGWAS database).</li>
  <li><code>Preclumped</code>: Whether or not the dataset was already [clumped](https://mr-dictionary.mrcieu.ac.uk/term/ld/) (e.g., data from OpenGWAS can be retrieved after undergoing clumping with the standard parameters).</li>
  <li><code>Rem.F.Stat</code>: Number of SNPs which were removed due to not meeting the [F-statistic](https://mr-dictionary.mrcieu.ac.uk/term/weak-instrument/) cut-off.</li>
  <li><code>Rem.Clumped</code>: Number of SNPs which were removed due to [clumping](https://mr-dictionary.mrcieu.ac.uk/term/ld/). Note that this will be 0 if the data was pre-clumped.</li>
  <li><code>SNPs.Formatted</code>: Number of SNPs which were successfully formatted to the MR-Base/TwoSampleMR package format (e.g., SNPs which had negative standard errors are removed at this stage).</li>
</ul>

## Exposure SNPs

This table details the final list of formatted SNPs for the exposure datasets which were used in the MR analyses.

```{r exp-dat}
params$dat[, colnames(params$dat) %in% c("SNP",
                                        "effect_allele.exposure",
                                        "other_allele.exposure",
                                        "beta.exposure",
                                        "eaf.exposure",
                                        "chr",
                                        "pos",
                                        "se.exposure",
                                        "pval.exposure",
                                        "exposure",
                                        "mr_keep.exposure",
                                        "f.stat.exposure")] %>%
  unique() %>%
  dplyr::rename(Eff.Allele = effect_allele.exposure,
                Alt.Allele = other_allele.exposure,
                Beta = beta.exposure,
                EAF = eaf.exposure,
                Chr = chr,
                Pos = pos,
                SE = se.exposure,
                P.value = pval.exposure,
                Exposure = exposure,
                Inc.MR = mr_keep.exposure,
                F.Stat = f.stat.exposure) %>%
  DT::datatable(rownames = F,
                caption = "Details for SNPs which were included in the MR analysis by passing the QC and cleaning stages.",
                extensions = c("Buttons"),
                options = list(order = list(8, "desc"), 
                               dom = "BDfrtip",
                               buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download')))) %>%
  DT::formatSignif(columns = c("P.value"), digits = 3) %>%
  DT::formatRound(columns = c("F.Stat"), digits = 0)
```

## Outcome datasets

There were `r nrow(unique(params$report$outcomes))` outcome datasets provided. Details are provided in the following table for each dataset, including details on how many SNPs were proxied.

```{r outcomes}
params$report$outcomes %>%
  dplyr::mutate(Trait.Annotated = paste0("<a href='./traits/", Trait.Annotated, ".html'>", Trait.Annotated, "</a>"),
                Trait.ID = ifelse(From.IEUGWASDB == 1, paste0("<a href='https://gwas.mrcieu.ac.uk/datasets/", Trait.ID, "/'>", Trait.ID, "</a>"), Trait.ID)) %>%
  dplyr::relocate(Trait.ID, .after = Trait.Annotated) %>%
  dplyr::select(-From.IEUGWASDB, -Trait.Name) %>%
  DT::datatable(rownames = F,
                escape = F,
                filter = 'top',
                extensions = 'Buttons',
                caption = "Breakdown for each outcome dataset.",
                options = list(dom = 'Bfrtip',                                
                               buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download'))))

```
**Column headers and meanings**
<ul>
  <li><code>Trait.Annotated</code>: Name of the trait - Links to that trait's results report.</li>
  <li><code>Trait.ID</code>: Trait ID - Links to the OpenGWAS DB, if applicable.</li>
  <li><code>SNPs</code>: Number of SNPs initially loaded from the GWAS (either from the file or from the OpenGWAS database).</li>
  <li><code>Num.Proxies</code>: Number of proxied SNPs. A list of all unique SNPs from all exposure datasets are used to extract outcome SNPs -- if a SNP is not found then a proxy SNP is used instead.</li>
  <li><code>SNPs.Formatted</code>: Number of SNPs which were successfully formatted to the MR-Base/TwoSampleMR package format (e.g., SNPs which had negative standard errors are removed at this stage).</li>
</ul>

## Outcome SNPs

This table details the final list of formatted SNPs for the outcome datasets which were used in the MR analyses.

```{r out-dat}
params$dat[, colnames(params$dat) %in% c("SNP",
                                        "effect_allele.outcome",
                                        "other_allele.outcome",
                                        "beta.outcome",
                                        "eaf.outcome",
                                        "chr",
                                        "pos",
                                        "se.outcome",
                                        "pval.outcome",
                                        "outcome",
                                        "proxy.outcome",
                                        "proxy_snp.outcome")] %>%
  dplyr::rename(Eff.Allele = effect_allele.outcome,
                Alt.Allele = other_allele.outcome,
                Beta = beta.outcome,
                EAF = eaf.outcome,
                Chr = chr,
                Pos = pos,
                SE = se.outcome,
                P.value = pval.outcome,
                Outcome = outcome,
                Proxy = proxy.outcome,
                Proxy.SNP = proxy_snp.outcome) %>%
  DT::datatable(rownames = F,
                caption = "Details for outcome SNPs, or their proxies, which were included in the MR analysis.",
                extensions = c("Buttons"),
                options = list(order = list(8, "asc"), 
                               dom = "BDfrtip", 
                               buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download')))) %>%
  DT::formatSignif(columns = c("P.value"), digits = 3)
```

# MR Results

MR results are presented here for the [Wald ratio](https://mr-dictionary.mrcieu.ac.uk/term/wald-ratio/) (WR) and [inverse variance weighted](https://mr-dictionary.mrcieu.ac.uk/term/ivw/) (IVW). The per-trait reports detail further MR methods, such as [MR-Egger](https://mr-dictionary.mrcieu.ac.uk/term/mr-egger/), but are omitted here for brevity.

There were a total of `r 0.05 / params$report$bonferroni` MR tests conducted. A Bonferroni P-value threshold would therefore be `r params$report$bonferroni` ($0.05 / tests$).

```{r mr, echo = F, message = F}
steiger <- params$report$steigerresults %>% 
  dplyr::select(id.exposure, id.outcome, flag)

params$report$results %>%
  dplyr::left_join(steiger) %>%
  dplyr::rename(Exposure.ID = id.exposure,
                Outcome.ID = id.outcome,
                Outcome = outcome,
                Exposure = exposure,
                Method = method,
                SNPs = nsnp,
                P.value = pval,
                OR = or,
                Lower.95ci = or_lci95,
                Upper.95ci = or_uci95,
                Steiger.Flag = flag) %>%
  dplyr::select(-Exposure.ID, -Outcome.ID) %>%
  dplyr::relocate(Outcome, .after = Exposure) %>%
  DT::datatable(rownames = F,
                filter = 'top',
                extensions = c("Buttons"),
                options = list(order = list(4, "asc"), 
                               dom = "Bfrtip", 
                               buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download')))) %>%
  DT::formatSignif(columns = c("P.value", "OR", "Lower.95ci", "Upper.95ci"), digits = 3)
```
**Column headers and meanings**
<ul>
  <li><code>Exposure</code>: Exposure trait names.</li>
  <li><code>Outcome</code>: Outcome trait names.</li>
  <li><code>Method</code>: MR method, either [WR](https://mr-dictionary.mrcieu.ac.uk/term/wald-ratio/) for single-SNP instruments or [IVW](https://mr-dictionary.mrcieu.ac.uk/term/ivw/) for multi-SNP instruments.</li>
  <li><code>SNPs</code>: Number of SNPs that form the instrument.</li>
  <li><code>P.value</code>: MR P value.</li>
  <li><code>OR</code>: MR odds ratios.</li>
  <li><code>Lower.95ci</code>: Lower 95% confidence interval.</li>
  <li><code>Upper.95ci</code>: Upper 95% confidence interval.</li>
  <li><code>Steiger.Flag</code>: Flag for the [Steiger filtering](https://mr-dictionary.mrcieu.ac.uk/term/steiger/) results: true, direction of effect is from exposure to the outcome; false, the direction of effect is from outcome to the exposure; unknown, Steiger P value is > 0.05.</li>
</ul>

# Colocalisation Results

Here are all of the [colocalisation](https://mr-dictionary.mrcieu.ac.uk/term/colocalization/) results. Regional plots are available in the per-trait reports.

```{r regional-plot, echo = F, message = F, include = F, dev = 'png', fig.show = 'hide'}
if (length(params$report$plots[params$report$plots$type == "regional", ]) > 0) {
  for (idx in as.integer(rownames(params$report$plots[params$report$plots$type == "regional", ])))
  {
    if (all(is.na(params$report$raw.plots[[idx]]))) {
      next
    }
    grid.newpage()
    grid.draw(params$report$raw.plots[[idx]])
  }
}
```

```{r coloc, echo = F, warning = F}
cresults <- params$report$cresults

for (i in 1:nrow(cresults)) {
    tryCatch({
      cresults[i, "plot"] <- sprintf("<a class='ItemsTooltip' target='_blank'><img class='imgTooltip' src='%s'/>Region</a>", knitr::image_uri(knitr::fig_chunk('regional-plot', 'png', i)))
    },
    error = function(e) {
      cresults[i, "plot"] <- "No plot"
    })
}
cresults <- cresults[cresults["nsnps"] > 50, ]

if (length(cresults)) {
  cresults %>%
    dplyr::rename(Exposure = name1,
                  Outcome = name2,
                  Plot = plot,
                  SNPs = nsnps,
                  ChrPos = chrpos) %>%
    dplyr::select(-id1, -id2) %>%
    dplyr::relocate(Plot, .after = Outcome) %>%
    DT::datatable(rownames = F,
                  escape = F,
                  filter = 'top',
                  extensions = c("Buttons"),
                  options = list(order = list(8, "desc"), 
                                 dom = "Bfrtip", 
                                 buttons = list('copy',
                                              list(extend = 'collection', 
                                                   buttons = c('csv', 'excel'), 
                                                   text = 'Download')))) %>%
    DT::formatPercentage(c("H0", "H1", "H2", "H3", "H4"), 2)
} else {
  cat("There are no colocalisation results.")
}
```
**Column headers and meanings**
<ul>
  <li><code>Exposure</code>: Name for Exposures.</li>
  <li><code>Outcome</code>: Name for Outcomes.</li>
  <li><code>Plot</code>: Hover over for the regional plot.</li>
  <li><code>SNPs</code>: Number of SNPs that form the instrument.</li>
  <li><code>H0-4</code>: Posterior probability for hypotheses 0 through 4.</li>
  <li><code>ChrPos</code>: Chromosome and position range used to extract SNPs for [colocalisation](https://mr-dictionary.mrcieu.ac.uk/term/colocalization/).</li>
</ul>
